pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,526 @@ package service
+	pods              atomic.Value
+	podMetrics        atomic.Value
+
+	jobConfigmapName string
+
+	statefulSets atomic.Value
+
+	leaderElectionID string
 
-func New(opts ...Option) (dsc Rebalancer, err error) {
+	rcd           time.Duration // reconcile check duration
+	eg            errgroup.Group
+	ctrl          k8s.Controller
+	tolerance     float64
+	rateThreshold float64
+	decoder       *decoder.Decoder
+}
+
+// NewRebalancer initialize job, configmap, pod, podMetrics, statefulset, replicaset, daemonset reconciler.
+// And it returns the rebalancer implemenation or any error occurred.
+func NewRebalancer(opts ...RebalancerOption) (Rebalancer, error) {
 	r := new(rebalancer)
-	for _, opt := range append(defaultOpts, opts...) {
+
+	for _, opt := range append(defaultRebalancerOpts, opts...) {
 		if err := opt(r); err != nil {
 			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
 		}
 	}
+
+	job, err := job.New(
+		job.WithControllerName("job rebalancer"),
+		job.WithNamespaces(r.jobNamespace),
+		job.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		job.WithOnReconcileFunc(func(jobList map[string][]job.Job) {
+			log.Debugf("[reconcile Job] length Joblist: %d", len(jobList))
+			jobs, ok := jobList[r.jobName]
+			if ok {
+				r.jobs.Store(jobs)
+			} else {
+				r.jobs.Store(make([]job.Job, 0))
+				log.Infof("job not found: %s", r.jobName)
+			}
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	cm, err := configmap.New(
+		configmap.WithControllerName("configmap rebalancer"),
+		configmap.WithNamespaces(r.jobNamespace),
+		configmap.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		configmap.WithOnReconcileFunc(func(configmapList map[string][]configmap.ConfigMap) {
+			configmaps, ok := configmapList[r.jobNamespace]
+			if ok {
+				for _, cm := range configmaps {
+					if cm.Name == r.jobConfigmapName {
+						if tmpl, ok := cm.Data[r.jobTemplateKey]; ok {
+							r.jobTemplate.Store(tmpl)
+						} else {
+							log.Infof("job template not found: %s", r.jobTemplateKey)
+						}
+						break
+					}
+				}
+			} else {
+				log.Infof("configmap not found: %s", r.jobConfigmapName)
+			}
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var rc k8s.ResourceController
+	switch r.agentResourceType {
+	case config.STATEFULSET:
+		rc, err = statefulset.New(
+			statefulset.WithControllerName("statefulset rebalancer"),
+			statefulset.WithNamespaces(r.agentNamespace),
+			statefulset.WithOnErrorFunc(func(err error) {
+				log.Error(err)
+			}),
+			statefulset.WithOnReconcileFunc(func(statefulSetList map[string][]statefulset.StatefulSet) {
+				log.Debugf("[reconcile StatefulSet] length StatefulSet[%s]: %d", r.agentName, len(statefulSetList))
+				sss, ok := statefulSetList[r.agentName]
+				if ok {
+					log.Debugf("[reconcile StatefulSet] StatefulSet[%s]: desired replica: %d, current replica: %d", r.agentName, *sss[0].Spec.Replicas, sss[0].Status.Replicas)
```

Comment:
sss[0] sometime makes panic please be careful

Reply:
should we check length before use it?

Reply:
I think so.

Reply:
thank you.
I fixed with the following commits.
https://github.com/vdaas/vald/pull/1187/commits/f7483e7534921921c548e82705c5b717fb88e7dc#diff-c9fe9660a4ab478b106a5fca677fb460f1f738eaff43f8037a356e3153018ff4R176